File: presences/src/main/resources/public/ts/services/CollectiveAbsenceService.ts
import {ng} from 'entcore';
import http, {AxiosResponse} from 'axios';
import {
    ICollectiveAbsence, ICollectiveAbsenceBody,
    ICollectiveAbsencesResponse,
    ICollectiveAbsenceStudent
} from '../models';

export interface CollectiveAbsenceService {

    getCollectiveAbsence(structureId: string, collectiveId: number): Promise<ICollectiveAbsence>;

    getCollectiveAbsences(structureId: string, params: ICollectiveAbsenceBody): Promise<ICollectiveAbsencesResponse>;

    createCollectiveAbsence(structureId: string, collectiveAbsence: ICollectiveAbsence): Promise<AxiosResponse>;

    updateCollectiveAbsence(structureId: string, collectiveAbsence: ICollectiveAbsence): Promise<AxiosResponse>;

    removeAbsenceFromCollectiveAbsence(structureId: string, collectiveId: number, studentIds: Array<string>): Promise<AxiosResponse>;

    deleteCollectiveAbsence(structureId: string, collectiveId: number): Promise<AxiosResponse>;

    exportCollectiveAbsences(structureId: string, startDate: string, endDate: string): void;

    getStudentsAbsencesStatus(structureId: string, params: ICollectiveAbsenceBody): Promise<Array<ICollectiveAbsenceStudent>>;
}

export const collectiveAbsenceService: CollectiveAbsenceService = {

    /**
     * Get collective absence from id.
     * @param structureId       structure identifier
     * @param collectiveId      collective absence identifier
     */
    getCollectiveAbsence : async (structureId: string, collectiveId: number): Promise<ICollectiveAbsence> => {

        return http.get(`/presences/structures/${structureId}/absences/collectives/${collectiveId}`)
            .then((res: AxiosResponse) => { return res.data; });
    },

    /**
     * Get list of collective absences.
     * @param structureId  structure identifier
     * @param params       collective absences filter parameters
     */
    getCollectiveAbsences : async (structureId: string, params: ICollectiveAbsenceBody): Promise<ICollectiveAbsencesResponse> => {
        let urlParams: string = `?startDate=${params.startDate}&endDate=${params.endDate}`;

        if (params.groups.length > 0) {
           for (let i = 0; i < params.groups.length; i++) {
               urlParams += `&audienceName=${params.groups[i]}`;
           }
        }

        urlParams += (params.page !== undefined && params.page !== null)  ? `&page=${params.page}` : '';

        return http.get(`/presences/structures/${structureId}/absences/collectives${urlParams}`)
            .then((res: AxiosResponse) => {
                return res.data;
            });
    },

    /**
     * Create collective absence.
     * @param structureId           structure identifier
     * @param collectiveAbsence     collective absence to create
     */
    createCollectiveAbsence : async (structureId: string, collectiveAbsence: ICollectiveAbsence): Promise<AxiosResponse> => {
        return http.post(`/presences/structures/${structureId}/absences/collectives`, collectiveAbsence);
    },

    /**
     * Update collective absence.
     * @param structureId           structure identifier
     * @param collectiveAbsence     collective absence to update
     */
    updateCollectiveAbsence : async (structureId: string, collectiveAbsence: ICollectiveAbsence): Promise<AxiosResponse> => {
        return http.put(`/presences/structures/${structureId}/absences/collectives/${collectiveAbsence.id}`, collectiveAbsence);
    },

    /**
     * Remove an absence from the collective absence.
     * @param structureId       structure identifier
     * @param collectiveId      collective absence identifier
     * @param studentIds        list of student identifiers
     */
    removeAbsenceFromCollectiveAbsence : async (structureId: string, collectiveId: number, studentIds: Array<string>): Promise<AxiosResponse> => {
        const students: {studentIds: Array<string>} = {
            studentIds: studentIds
        };

        return http.put(`/presences/structures/${structureId}/absences/collectives/${collectiveId}/students`, students);
    },

    /**
     * Delete collective absence.
     * @param structureId       structure identifier
     * @param collectiveId      collective absence identifier
     */
    deleteCollectiveAbsence : async (structureId: string, collectiveId: number): Promise<AxiosResponse> => {
        return http.delete(`/presences/structures/${structureId}/absences/collectives/${collectiveId}`);
    },

    /**
     * Export collective absences in CSV file.
     * @param structureId       structure identifier
     * @param startDate         start date range
     * @param endDate           end date range
     */
    exportCollectiveAbsences : (structureId: string, startDate: string, endDate: string): void => {
        window.open(`/presences/structures/${structureId}/absences/collectives/export?startDate=${startDate}&endDate=${endDate}`);
    },

    /**
     * Get list of student absences status
     * @param structureId       structure identifier
     * @param body              collective absences parameters
     */
    getStudentsAbsencesStatus : async (structureId: string, body: ICollectiveAbsenceBody): Promise<Array<ICollectiveAbsenceStudent>> => {
        return http.post(`/presences/structures/${structureId}/absences/isAbsent`, body)
            .then((res: AxiosResponse) => { return res.data.all; });
    }
};

export const CollectiveAbsenceService = ng.service('CollectiveAbsenceService',
    (): CollectiveAbsenceService => collectiveAbsenceService);

File: presences/src/main/java/fr/openent/presences/service/CollectiveAbsenceService.java
package fr.openent.presences.service;

import fr.wseduc.webutils.Either;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import org.entcore.common.user.UserInfos;

import java.util.List;

public interface CollectiveAbsenceService {

    /**
     * get absences
     *
     * @param structureId Structure identifier
     * @param startDate   startDate
     * @param endDate     endDate
     * @param handler     Function handler returning data
     */
    void getCollectives(String structureId, String startDate, String endDate, Long reasonId, Boolean regularized,
                        List<String> audienceNames, Integer page, Handler<AsyncResult<JsonObject>> handler);

    void get(String structureId, Long collectiveId, Handler<AsyncResult<JsonObject>> handler);

    void getCollectiveFromAbsence(Long absenceId, Handler<AsyncResult<JsonObject>> handler);

    void getAbsencesStatus(String structureId, List<String> studentIds, String startDate, String endDate,
                           Long collectiveId, Handler<AsyncResult<JsonObject>> handler);

    /**
     * create absence
     *
     * @param collectiveBody collectiveBody fetched
     * @param user        userInfo
     * @param handler     Function handler returning data
     */
    void create(JsonObject collectiveBody, UserInfos user, String structureId, Handler<AsyncResult<JsonObject>> handler);

    /**
     * update absence
     *
     * @param handler     Function handler returning data
     */
    void update(JsonObject collectiveBody, UserInfos user, String structureId, Long collectiveId, Handler<AsyncResult<JsonObject>> handler);


    /**
     * Delete absence from collective absence.
     * @param studentIds    list of student identifiers
     * @param structureId   structure identifier
     * @param collectiveId  collective absence identifier
     * @param handler       Function handler returning data
     */
    void removeAbsenceFromCollectiveAbsence(JsonObject studentIds, String structureId, Long collectiveId, Handler<AsyncResult<JsonObject>> handler);

    /**
     *
     * @param structureId   structure identifier
     * @param collectiveId  collective identifier
     * @param handler       Function handler returning data
     */
    void removeAudiencesRelation(String structureId, Long collectiveId, Handler<AsyncResult<JsonObject>> handler);

    /**
     * delete absence
     *
     * @param handler   Function handler returning data
     */
    void delete(Long id, String structureId, Handler<AsyncResult<JsonObject>> handler);


    /**
     * Retrieve collective absences for CSV export
     *
     * @param structureId   structure identifier
     * @param startDate     start date range
     * @param endDate       end date range
     * @param handler       Function handler returning data. Returns a JsonArray
     */
    void getCSV(String structureId, String startDate, String endDate, Handler<AsyncResult<JsonArray>> handler);
}
